BUGFIX: Simplify PartialPublish & Discard (for now) to NOT select dimensions and have custom events

mhsdesign opened the pull request:
Partially fixes: #4614 (as we would publish the node move that is marked as moved in some random dimension)
Partially fixes: https://github.com/neos/neos-development-collection/issues/5387
See also slack: https://neos-project.slack.com/archives/C04PYL8H3/p1732899328791309

### Why the `NodeIdentifiersToPublishOrDiscard` is removed from the event

Only with https://github.com/neos/neos-development-collection/commit/f2c66e08c92be23ad0b860547c5d3c1b46210a66
the `NodeIdentifiersToPublishOrDiscard` was introduced to the `WorkspaceWasPartiallyPublished` and `DiscardIndividualNodesFromWorkspace` events. Previously they did not contain the information which "filter" was used to create the events. And strictly speaking - for the ESCR core content graph this information on the event is purely cosmetic.

The only use currently is the asset usage catchup hook that uses the node ids and their dimension if a workspace was discarded to remove usage data:

https://github.com/neos/neos-development-collection/blob/be1f1fe6f430ae2d44765c64d3492fe7f2dddfb4/Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php#L62

One could also optimise the content cache flusher for the partial discard case to not flush the whole workspace but only partial entries:

https://github.com/neos/neos-development-collection/pull/5160/files#diff-008f3460b749dbf869e63ba0602a1501990fa09cc37cd7fdaca6afa3789017e4R143

As we currently flush the whole workspace on discard:

https://github.com/neos/neos-development-collection/blob/902e08d249c7a5145b1b9522efc6d0f86d072cdc/Neos.Neos/Classes/Fusion/Cache/GraphProjectorCatchUpHookForCacheFlushing.php#L196

But this is not worth to optimise as we discussed: https://github.com/neos/neos-development-collection/issues/5175#issuecomment-2386698992

So as this is the only - discussable use-case - we dont want to make the event creation more complicated. Instead of gathering the information which nodes were really affected and which dimensions we just fully leave out this information in the event. (Like we do also with other events where we don't specify which nodes were all actually removed in NodeAggregateWasRemoved but rely on that the projection does what it needs).

First i tried to extract the information for the `$publishedNodes` and `$discardedNodes` as shown in this pr in `extractNodeAggregateIdsAndAffectedDimensionSpacePoints` https://github.com/neos/neos-development-collection/pull/5386 but its problems are further described there.

Further treating a partial and full discard the same also simplifies the catchup hooks.


### Why the distinction between full and partial publish is removed from the event

Without the need to specify the `NodeIdentifiersToPublishOrDiscard` (see above) there is no reason to differentiate between the events at all.

Thus we upcast the the partial publish and discard events to a publish or discard inside `fromArray`:

```php
public static function fromArray(array $values): EventInterface
{
    return new WorkspaceWasPublished(
        WorkspaceName::fromString($values['sourceWorkspaceName']),
        WorkspaceName::fromString($values['targetWorkspaceName']),
        ContentStreamId::fromString($values['newSourceContentStreamId']),
        ContentStreamId::fromString($values['previousSourceContentStreamId']),
    );
}
```

Both events share the same properties their counter parts have.
Well keep the `$publishedNodes` and `$discardedNodes` fields in the database as they don't do harm.

`WorkspaceWasPartiallyPublished` -> `WorkspaceWasPublished`

- `WorkspaceName $sourceWorkspaceName`
- `WorkspaceName $targetWorkspaceName`
- `ContentStreamId $newSourceContentStreamId`
- `ContentStreamId $previousSourceContentStreamId`
- ignore: `NodeIdsToPublishOrDiscard $publishedNodes`

`WorkspaceWasPartiallyDiscarded` -> `WorkspaceWasDiscarded`

- `WorkspaceName $workspaceName`
- `ContentStreamId $newContentStreamId`
- `ContentStreamId $previousContentStreamId`
- ignore: `NodeIdsToPublishOrDiscard $discardedNodes`

To allow to distinct what events are part of a (partial) publish or not, instead the causation id will be used in the future: https://github.com/neos/neos-development-collection/issues/3887 (to be implemented in a followup)


**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

TODO:
- [x] adjust tests that test that partial publish is a full publish
- [x] check against left todos!


**Review instructions**

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
TASK: Remove possibility to specify dimensions in partial publish or delete

mhsdesign committed:
TASK: Remove obsolete `MatchableWithNodeIdToPublishOrDiscardInterface`

the logic is carried out inside `RebaseableCommands::commandMatchesAtLeastOneNode`

The reasoning behind this is that asking the commands which node aggregate id is affected _does_ work but the capabilities stop if we want to ask which dimensions are affected.

For that we must use the event instead - at least in the future if we want to reintroduce dimension based publishing again. The matching list from commands to node id will also be simplified once we use the `EmbedsNodeAggregateId` interface instead and operate on events.

mhsdesign committed:
TASK: Improve asset usage test match error

kitsunet commented:
Still don't like it, but probably best way forward for now.

mhsdesign committed:
TASK: Revert catchup hook adjustment to rely on `NodeAggregateIds` instead of `NodeIdsToPublishOrDiscard`

... as a 3rd option the followup commit will just flush the full asset usage workspace!

mhsdesign committed:
TASK: Remove information which nodes were partially published or discarded in the events (for now)

Introduced with https://github.com/neos/neos-development-collection/commit/f2c66e08c92be23ad0b860547c5d3c1b46210a66 - its not important for the graph projection and only used by the asset usage catchup hook which can be adjusted.

If we want to re-introduce `$publishedNodes` to the event should be also careful that we calculate them from the _actual_ events so we can know for sure there is no gibberish in the event.
Previously for example discarding a non-existing node or a node in a not valid DSP would just save that directly into the event, which is not correct as that did not happen like that. It was merely a bad intention.

mhsdesign committed:
TASK: Simplify asset usage catchup hook to not optimise partial discard

Instead of fine granular updating only the usage references on the nodes that were discarded, well throw away everything and then rely on the reapplied events what is kept.

This is more consistent to how other parts react to publishing like the graph just forks a new content stream and doesnt care about the old data, the change projection behaves similar to the graph.
And for the content cache flusher we had the discussion to use `discardedNodes` but opted against it and do a full flush there as well: https://github.com/neos/neos-development-collection/issues/5175#issuecomment-2386698992

This change allows us to get rid of the `discardedNodes` property (with dimensions) in the `WorkspaceWasPartiallyDiscarded` event.

mhsdesign committed:
WIP: Simplify publishing, that partial and full publish is the same

reasoning:

- the `publishedNodeIds` were removed on the event, as we dont need it, and it would rise a discussion if we need to keep dimension space points there too or not: see https://github.com/neos/neos-development-collection/pull/5386 how i archived it.

- instead of now trying to make `WorkspaceWasPartiallyPublished` contain all the information we dont use or need, the idea is to make it dumb. Now its basically just a `WorkspaceWasPublished` and then the other remaining events are applied as if they were carried out by a user.
If we need to add more metadata to the partial publish at some point, that would need a new event id say as we woouldnt be able to mess with the `WorkspaceWasPartiallyPublished` once released.
So by just combining them into one we have the option open to create a `WorkspaceWasPartiallyPublished` again but differently which wil be breaking too but differently. Its just a new event then, no need to migrate data.

mhsdesign commented:
Soooo i stripped this pr a little futher to its needed parts. And its beautiful. +130 lines -400lines. With the last controversial commit another -200 lines.
So a lot complexity is gone ^^ the parts that were perviously in here - extracting the node ids and their affected dsp as metadata for the event - are now part of https://github.com/neos/neos-development-collection/pull/5386
This 'feature' is unused (expect on place in the asset usage) and fully untested in the core. 

Adjustment in the asset usage:

> [TASK: Simplify asset usage catchup hook to not optimise partial discard](https://github.com/neos/neos-development-collection/pull/5385/commits/e4ad10b4484bd9d8357b2c17e33a25538778a365)
> Instead of fine granular updating only the usage references on the nodes that were discarded, well throw away everything and then rely on the reapplied events what is kept.
>
> This is more consistent to how other parts react to publishing like the graph just forks a new content stream and doesnt care about the old data, the change projection behaves similar to the graph.
> And for the content cache flusher we had the discussion to use `discardedNodes` but opted against it and do a full flush there as well: https://github.com/neos/neos-development-collection/issues/5175#issuecomment-2386698992
> 
> This change allows us to get rid of the `discardedNodes` property (with dimensions) in the `WorkspaceWasPartiallyDiscarded` event.

mhsdesign committed:
Linter in the winter

mhsdesign committed:
TASK: Introduce migration to rename WorkspaceWasPartiallyPublished and WorkspaceWasPartiallyDiscarded

mhsdesign committed:
TASK: Phpstanisize event migrations

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into bugfix/publish-only-across-all-dimensions

mhsdesign committed:
TASK: Adjust new change projection tests to dont handle discarding one dimension for now

dlubitz commented on Neos.ContentRepository.Core/Classes/Feature/RebaseableCommands.php:
```diff
@@ -67,11 +72,37 @@ public function separateMatchingAndRemainingCommands(
-            if ($command->matchesNodeId($nodeId)) {
+        foreach ($nodeIdsToMatch as $nodeId) {
+            /**
+             * This match must contain all commands which are working with individual nodes, such that they are
+             * filterable whether they are applying their action to a NodeIdToPublish.
```
I don't get what this means. What does "whether they are applying their action to a NodeIdToPublish" mean? AFAIS we have no "NodeIdToPublish" anymore.

dlubitz commented:
I don't understand, why we need to remove the "WasPartially*" Events. This is exactly what we shouldn't do in a event sourced system. I see the point, that we don't need to handle that anymore, but this shouldn't be the way to go.

mhsdesign commented:
Its not the classy thing but were beyond that - especially as its beta id say. The point why i exactly removed it are:

- a replay is NOT necessary and the migration is only needed for if you want to replay some time - so the current state of the db is fine
- at some point we _might_ reintroduce the WasPartially events with proper metdata and information, and to avoid having WasPartially_2 then its just easier to get rid of them now as we dont need them
- its easy to forget implementing one of the events like PublishWorkspace or the partial thing, to make it simpler to write catchup hooks its just sane to have one event

i also discussed this approach with christian and Sebastian and thing its the right one? - also in terms of reducing complexity for once.

kitsunet reviewed (approved):
Yeah as I said before I am not happy at all with this, but it's probably safer for now. The "lost" information is minimal, I don't see a scenario in which we retroactively would want, need to do something based on the partially (published|discarded) events.

mhsdesign committed:
TASK: Revert special code path from partial to full publish

Reverts d2d9faa0b46a9da323ed059d7309b7b1247b95fa partially

Initially this change was done to fix #5303
But that was a hoax see https://github.com/neos/neos-development-collection/issues/5303#issuecomment-2528984614

we always emit the `WorkspaceWasPublished` event and this just simplifies the code-path again.

For the discard all case, the code is stil a little easier to read _with_ the condition.

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/RebaseableCommands.php:
```diff
@@ -67,11 +72,37 @@ public function separateMatchingAndRemainingCommands(
+                )?->equals($nodeId),
+                SetSerializedNodeReferences::class => $command->sourceNodeAggregateId->equals($nodeId),
+                // todo return false instead to allow change to be still kept?
+                default => throw new \RuntimeException(sprintf('Command %s cannot be matched against a node aggregate id. Partial publish not possible.', $command::class), 1645393655)
```
Todo - needs test - there is a bug in partial publish and discard which means that if there is a command on the workspace NOT in this list, no partial operation is allowed. This is a bit to hard. Instead simply keeping that change as remainder makes more sense imo?

mhsdesign replied:
Solved via https://github.com/neos/neos-development-collection/pull/5385/commits/fa58f0c0f2f59cc41a1864ba6808e126f2677fd5

mhsdesign committed:
TASK: Remove obsolete todos

mhsdesign committed:
TASK: Remove migration for partially publish events and handle them at runtime

We cannot erase history just because we want to.
Especially because this does not fix any bugs.

mhsdesign committed:
TASK: Simplify upcasting of legacy events to be in `fromArray` instead spread around

This overhauls the `NodeAggregateWasDisabled` upcasting from

https://github.com/neos/neos-development-collection/pull/4659

mhsdesign commented:
Okay i give in. @dlubitz youre right we dont need to rewrite history if we dont need to. i relaxed the `fromArray` to allow to return other objects than `self`. Now its happening at runtime.

- [TASK: Remove migration for partially publish events and handle them a…](https://github.com/neos/neos-development-collection/pull/5385/commits/3412594387fa436d108affa3ed960db832964680) 
- [TASK: Simplify upcasting of legacy events to be in fromArray instea…](https://github.com/neos/neos-development-collection/pull/5385/commits/135dc67e89bdf90f6cc88c71b76de34001ab461d)

Also the content stream pruner will still select the legacy events in the query having in mind, that they will be upcasted.

bwaidelich commented:
I'm late to the party and only skimmed through the code so far.
I fail to understand why those `Partially` events ought to be removed.
IMO there's a crucial domain-specific difference between a full and a partial publish/discard. And this difference should not only be reflected by some metadata in the events.
But most probably I misunderstood things

mhsdesign committed:
TASK: Introduce "partial" flag on publish events

This was done do ease debugging and still preserve the intention.

mhsdesign commented:
As discussed today, i added a `partial` flag to the events as we want keep the main idea to reduce the complexity (only one public event)

See: [TASK: Introduce "partial" flag on publish events](https://github.com/neos/neos-development-collection/pull/5385/commits/6bc3c55654b87c9d126f1e70a83bbbb00bdb6c59)

mhsdesign committed:
BUGFIX: Ensure that `commandMatchesAtLeastOneNode` contains all commands for match

kitsunet commented:
I guess we can merge this then?
